### PR TITLE
Configuration required to potentially have datanode task on namenode or journal node

### PR DESCRIPTION
**Pull request by nigoel**

### CODE REVIEW

**Comment by elingg** [comment]
thanks @nigoel! The concern here is that it adds more complexity to our configuration although it will allow us to run on fewer machines. @brndnmtthws, in your experience with production HDFS ops, were there issues with I/O related to running datanodes on the namenode machines?

**Comment by nigoel** [comment]
Hi elingg, For the same, I have kept the default in code to keep the behavior same. In my case, I had an exclusive set of compute machines (by attributes in mesos) on which I wanted to run my particular job. That job had some data sandboxing issues, so required a separate hdfs through resource reservations. Making it a requirement to ave 3+ node, makes a "less resource utilization" solution to me. Reason for choosing the mesos was to have optimal resource utilization. Since ultimately mesos will run everything in linux containers (typical installation of mesos has containerization on), it should not matter whether we actually run them on different physical boxes or same.

**Comment by brndnmtthws** [comment]
@elingg It's probably unwise to run datanodes and namenodes together.

Furthermore, separate services *should* write to separate volumes to avoid OS disk buffer issues.

**Comment by nigoel** [comment]
@elingg @brndnmtthws Are there some architectural issues with this specific implementation of hdfs on Mesos. Because normal installation of hdfs don't otherwise restrict it. 

For example, this public document provides the default documentation where DN and NN are on same node. https://hadoop.apache.org/docs/r2.4.1/hadoop-project-dist/hadoop-hdfs/HdfsUserGuide.html 

Mesos is in fact one step more safe from the perspective of resources rather DN and NN on same node on normal cluster, because there will be no memory, disk and CPU contentions between DN and NN tasks. In Mesos, architecturally DS nodes will horizontally scale. If I have high end cluster machines, why should I waste resources and order more machines.

**Comment by elingg** [comment]
I understand Mesos does have great isolation, but there is still room to go. There will still be high IO usage/ Disk Buffers issues especially since we haven't made use of persistent volumes as of yet. Another issue is with increasing the possibility of failures. I would love more input here though @brndnmtthws, @kensipe, @gabrielhartmann, please let me know your input

**Comment by nigoel** [comment]
@elingg  I was about to come on that. Usage of persistent volume will solve that problem. It is very unfair to assume that world is going to use this implementation of yours in a set-up that mesosphere has designed. For example, we in our project are working on using the persistent volumes and that's the problem we need to fix while running the docker or our cluster provisioning. Not everything meant to be fixed in hdfs-mesos. Since people using it will understand the repercussions, we should not strictly make it unusable for people who had already solved the issue with their own definition of containerizers or their custom executors that ensures higher level of sandboxing.  Having said that, I strongly believe default setting should not allow this. Same is reflected in PR.

**Comment by brndnmtthws** [comment]
I'd prefer to be conservative, and err on the side of reducing risk associated. My 2 cents.

**Comment by elingg** [comment]
I'm reluctant to merge this change, especially pre persistent volumes, for the reason @brndnmtthws mentioned.

**Comment by elingg** [comment]
Maybe we could add a comment "Use at your own risk" in order to let people know the associated risks

**Comment by nigoel** [comment]
@elingg  I went to persistent volume docs on mesos. Even that doesn't solve the case. It can just ensure that you get "your persisted id" resource back where you have stored some data. But there is no isolation in terms of disk buffers. This can only be achievable by creating a virtual environment. 

Also, I am not sure what strategy normal hdfs installation (without mesos) have to prevent this. I believe it doesn't.

**Comment by elingg** [comment]
@nigoel, yes you are correct that the first version of persistent volumes does not prevent these issues. I'm pretty sure that this design will evolve to increase isolation since that is what we do with memory and cpu in the Mesos containerizer-why not disk. Normal HDFS installation does not prevent this, as you mentioned, it is up to the operator to install on the appropriate machines.

I am still considering if we should make this option configurable and "use at your own risk" or whether we should avoid this risk and require the user to deploy on more machines. At the moment, I'd still like to err on the side of caution. Community input appreciated...

**Comment by nigoel** [comment]
I have update the warnings and moved the configuration to commented tags

**Comment by keithchambers** [comment]
@elingg I recommend we merge this.  It's not a huge code change and we have necessary warnings in place to prevent production users from unknowingly shooting themselves in the foot.

File: hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/DataNode.java [hunk 1]
```diff
@@ -32,10 +32,12 @@ public boolean evaluate(Offer offer) {
       // What number of DN's should we try to recover or should we remove this constraint
       // entirely?
       if (deadDataNodes.isEmpty()) {
-        if (persistenceStore.dataNodeRunningOnSlave(offer.getHostname())
-            || persistenceStore.nameNodeRunningOnSlave(offer.getHostname())
-            || persistenceStore.journalNodeRunningOnSlave(offer.getHostname())) {
-          log.info(String.format("Already running hdfs task on %s", offer.getHostname()));
+        if (persistenceStore.dataNodeRunningOnSlave(offer.getHostname())) {
+          log.info(String.format("Already running datanode hdfs task on %s", offer.getHostname()));
```
Comment (elingg) [inline hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/DataNode.java]: nit don't add a space before config.getRunDatanodeExclusively()

**Comment by elingg** [comment]
Thanks @nigoel for adding the warning. This is a small code and change and should be fine to merge with that warning. Could you kindly rebase with master and then we can merge? Thanks!

File: hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/DataNode.java [hunk 2]
```diff
@@ -32,10 +32,12 @@ public boolean evaluate(Offer offer) {
...
+        } else if ( config.getRunDatanodeExclusively() && 
+              (persistenceStore.nameNodeRunningOnSlave(offer.getHostname())
+             || persistenceStore.journalNodeRunningOnSlave(offer.getHostname()))) {
```
Comment (elingg) [inline hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/DataNode.java]: nit: say Already running namenode or journalnode hdfs task on %s

**Comment by nigoel** [comment]
Thanks @elingg. Done.

**Pull request merged**